Source/WebCore/rendering/GridMasonryLayout.cpp

```diff
@@ -0,0 +1,143 @@
+{
+    m_masonryAxisDirection = masonryAxisDirection;
+    m_gridAxisDirection = m_masonryAxisDirection == ForRows ? ForColumns : ForRows;
+    m_masonryAxisGridGap = m_renderGrid.gridGap(m_masonryAxisDirection);
+    m_gridAxisTracksCount = m_renderGrid.numTracks(m_gridAxisDirection);
+    m_runningPositions = Vector<LayoutUnit>(m_gridAxisTracksCount);
+
+    // 2.4 Masonry Layout Algorithm
+    addItemsToFirstTrack(firstTrackItems);
+    placeItemsWithDefiniteGridAxisPosition(itemsWithDefiniteGridAxisPosition);
+    placeItemsWithIndefiniteGridAxisPosition(itemsWithIndefinitePosition);
+}
+
+void GridMasonryLayout::addItemsToFirstTrack(const HashMap<RenderBox*, GridArea>& firstTrackItems)
+{
+    for (auto& itemItr : firstTrackItems) {
+        auto* item = itemItr.key;
+        auto gridArea = itemItr.value;
+
+        insertIntoGridAndLayoutItem(item, gridArea);
+    }
+}
+
+void GridMasonryLayout::placeItemsWithDefiniteGridAxisPosition(const Vector<RenderBox*>& itemsWithDefiniteGridAxisPosition)
+{
+    for (auto* item : itemsWithDefiniteGridAxisPosition) {
+        auto itemSpan = m_renderGrid.currentGrid().gridItemSpan(*item, m_gridAxisDirection);
+
+        ASSERT(!itemSpan.isIndefinite());
+
+        auto initialTrackSpan = GridSpan::translatedDefiniteGridSpan(0, 1);
+        itemSpan.translate(m_renderGrid.currentGrid().explicitGridStart(m_gridAxisDirection));
+
+        GridArea area = m_masonryAxisDirection == ForRows ? GridArea(initialTrackSpan, itemSpan) : GridArea(itemSpan, initialTrackSpan);
+        insertIntoGridAndLayoutItem(item, area);
+    }
+}
+
+void GridMasonryLayout::placeItemsWithIndefiniteGridAxisPosition(const Vector<RenderBox*>& itemsWithIndefinitePosition)
+{
+    for (auto* item : itemsWithIndefinitePosition)
+        insertIntoGridAndLayoutItem(item, nextMasonryPositionForItem(item));
+}
+
+void GridMasonryLayout::insertIntoGridAndLayoutItem(RenderBox* child, const GridArea& area)
+{
+    m_renderGrid.currentGrid().insert(*child, area);
+
+    if (m_gridAxisDirection == ForColumns)
+        child->setOverridingContainingBlockContentLogicalWidth(m_renderGrid.m_trackSizingAlgorithm.estimatedGridAreaBreadthForChild(*child, ForColumns));
+    else
+        child->setOverridingContainingBlockContentLogicalHeight(m_renderGrid.m_trackSizingAlgorithm.estimatedGridAreaBreadthForChild(*child, ForRows));
+    child->layoutIfNeeded();
+    updateRunningPositions(child, area);
+}
+
+void GridMasonryLayout::updateRunningPositions(const RenderBox* child, const GridArea& area)
+{
+    auto gridSpan = m_masonryAxisDirection == ForRows ? area.columns : area.rows;
+
+    ASSERT(gridSpan.startLine() < m_runningPositions.size() && gridSpan.endLine() <= m_runningPositions.size());
+    gridSpan.clamp(m_runningPositions.size());
+
+    auto previousRunningPosition = m_runningPositions[gridSpan.startLine()];
```

Comment:
For non-'first track' items, shouldn't this iterate all the lines in `gridSpan` and pick the largest?

Reply:
Yup this should be fixed in https://github.com/WebKit/WebKit/pull/6940